utils/testrunner/TestRunner.js

```diff
@@ -335,6 +334,66 @@ class TestRunner {
     this._workers = [];
     this._terminating = false;
     this._result = null;
+    this._environmentToGlobalState = new Map();
+  }
+
+  async _willStartTestRun(worker, testRun) {
+    testRun._startTimestamp = Date.now();
+    testRun._workerId = worker._workerId;
+    await this._runDelegateCallback(this._delegate.onTestRunStarted, [testRun]);
+  }
+
+  async _didFinishTestRun(worker, testRun) {
+    testRun._endTimestamp = Date.now();
+    testRun._workerId = worker._workerId;
+    for (const environment of testRun._environments) {
+      const globalState = this._environmentToGlobalState.get(environment);
+      globalState.pendingTestRuns.delete(testRun);
+    }
+
+    await this._runDelegateCallback(this._delegate.onTestRunFinished, [testRun]);
+  }
+
+  async _runGlobalSetup(worker, environment) {
+    const globalState = this._environmentToGlobalState.get(environment);
+    if (!globalState.globalSetupPromise) {
+      globalState.globalSetupPromise = (async () => {
+        let result = true;
+        for (const hook of environment.hooks('globalSetup')) {
+          if (!await worker._runHook(null, hook, environment.name(), []))
```

Comment:
It feels strange that we use `worker._runHook` here. Should we introduce a `HookRunner` that all workers and test runner itself can use?